### ISSUE TITLE
Accessibility

### ISSUE THREAD

**Issue by ceciliaconsta3** [freeCodeCamp/chapter#65]
# Keeping accessibility in mind

As we are creating a new application that brings together a variety of people without a paywall *cough Wework cough*, we need to keep accessibility in mind.  Having the application being at least AA WCAG compliant would ensure we can at least reach users with some level of visual or physical impairment. By adding minimal tweaks here and there, we can reach more by removing barriers to the platform, hence "accessibility". 

Since FCC is not owned or used federally, we can skip Section 508 but I added it below for reference if anyone is interested in seeing where it all started. If anyone has additional ideas on implementation or other helpful comments, please do share and feel free to make changes as needed.

## A Practical Approach
- **Color Contrast:** Maybe white text on a yellow background looks cute but if people can't read it...
- **Use Semantic markup:** use HTML5 tags as they were intended, use CSS for the rest
- **Screen reader:** Use this chrome extension or squint and see if you can still find what you need
- **Aria Labels, tooltips, titles, and alt properties**: making sure images have fallback text and if they contain text, make sure important content is not lost
- **Skip to Content button:** avoid the user having to listen to every dropdown option in the navigation
- **Zoom:** Some users need to enlarge content, making sure things don't fall to pieces when they do
- **Font size:** 16px is the web standard. Let's save the light grey 10px font on a white background for things we intentionally don't want users to see

### Tools
[Axe Accessibility Chrome extension](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US)
[Function Accessibility Evaluator](https://fae.disability.illinois.edu/anonymous/?Anonymous%20Report=/)
[Web Aim Color Contrast Checker](https://webaim.org/resources/contrastchecker/)
[ChromeLens chrome extension - seeing through another's eyes](https://chrome.google.com/webstore/detail/chromelens/idikgljglpfilbhaboonnpnnincjhjkd)
[Screen Readers: Narrator, AudioVox, NVDA, JAWS - prepare to be overwhelmed](https://soap.stanford.edu/tips/screen-reader-testing)

> Web Accessibility does not get enough coverage in CS or Web Dev courses, but it makes a huge difference for a lot of users online. - Roman Mikhailov

### References
[Overview (with videos)](https://medium.com/front-end-weekly/all-about-web-accessibility-304e2c814751)
[W3 Developing Tips](https://www.w3.org/WAI/tips/developing/)
[Google Overview on Design and UX - specifically ARIA labels](https://developers.google.com/web/fundamentals/accessibility)
[WCAG Quick Reference](https://www.w3.org/WAI/WCAG21/quickref/#non-text-content)
[Quick tips using manual testing](https://dev.to/lkopacz/4-things-that-i-always-manually-test-2je8?utm_source=additional_box&utm_medium=internal&utm_campaign=regular&booster_org=)

## Extra
[Section 508 amendment details ](https://www.section508.gov/manage/laws-and-policies)

**Comment by chrismgonzalez** [comment]
This is awesome!  When we decide on a spec for this I (or someone) will add this to the docs.

**Comment by Critchkn** [comment]
Should probably also need forms to follow a logical tab order, and ensure the input fields have the proper aria role

**Comment by ryandudek** [comment]
I think we should advocate for using actual screen readers for testing as well as ChromeLens. On a Windows machine, you can download NVDA for free. On Mac, you can use VoiceOver - which is built in to your machine. This is much more accurate to what a user will end up with.

Also, can we add the [Toolness Accessible Color Matrix](https://toolness.github.io/accessible-color-matrix/) to the resources? It's pretty handy for showing what colors will and will not work together for AA compliance.

Also, I would love to contribute to the accessibility effort here.

**Comment by ceciliaconsta3** [comment]
I listed an article that referenced other screen readers but explicitly added those two under the Practical section + added your @Critchkn note on tab indexes for forms. Also @ryandudek , that Color Matrix is banging, added as well.

**Comment by allella** [comment]
Can we cross-reference the issues this is most likely to impact by # in'g them here, or # 'ing this on those threads?  #5 has accessibility in the first post, so it seems that one is related at a minimum.  Not sure how this would apply to mobile clients, and that could be further off, but maybe reference this to the mobile client thread too?

**Comment by allella** [comment]
Moving the list of accessibility interested folks here from the

- @timi-codes
- ceciliaconsta3 on Discord
- Critchkn on Discord
- @ryandudek
- you / @maiya-22
- codedawg on Discord
- @QuincyLarson said ["accessiblity is the law now"](https://discordapp.com/channels/633401573002969128/633401573002969132/633499775001690122)" so he's onboard

**Comment by Cheukting** [comment]
Thanks @ceciliaconsta3 it is a great suggestion. Having accessibility functions is definitely needed for a inclusive community app.

**Comment by maiya-22** [comment]
Maybe we can make a check-list that all new features need to be referenced against, to make sure that they are compliant with accessibility standards.

And we can comment about how they were made compliant, so that the community can learn about this essential skill, as we go.

**Comment by ceciliaconsta3** [comment]
@maiya-22 That's a great idea. A checklist or design guide is even more practical. I'll keep an eye out for a pre-made checklist online, possibly within one of the referenced sources, or create a curated list to add to the issue.

**Comment by Critchkn** [comment]
Is there anything I can do to help? I'll have some free time tomorrow and sunday

**Comment by ceciliaconsta3** [comment]
@Critchkn  Don't think so at the moment. I propose with every front end PR someone can volunteer as designated "accessibility checker". So for now, maybe adding yourself as an additional peer reviewer would help.

**Comment by shirakaba** [comment]
I've been advocating to use React Native Web for producing the web client, as it allows us to share a single React Native codebase to produce both a mobile client (#20) and a web client (not to mention other potential clients like desktop clients). I'd like to advocate it as a solution for cross-platform accessibility, too.

React Native includes [accessibility features](https://facebook.github.io/react-native/docs/accessibility.html), developed [by the Amazon team](https://github.com/facebook/react-native/pull/24095), so they should be industry-standard. By implementing accessibility in a React Native codebase, it would be inherited across all other platforms that React Native ports exist for, provided the port is thorough. React Native Web [implements accessibility thoroughly](https://github.com/necolas/react-native-web/blob/master/docs/guides/accessibility.md), and it powers the Twitter website (which purports to be accessible). [React Native Windows](https://github.com/Microsoft/react-native-windows) and [React Native macOS](https://github.com/Microsoft/react-native) are both developed by Microsoft and used in Office and other products, so I think that accessibility won't have been neglected for those ports, either.

I'd be interested in what an accessibility expert's opinion is on the Twitter website and native mobile apps. If both seem adequately accessible, it would be a mark of confidence for a client stack involving React Native and React Native Web.

**Comment by allella** [comment]
#144 could use input from the Accessibility-focused folks in this thread on where to draw the line on browser support. I'm making the assumption that socio-economic situation, and therefore available computers / devices, falls within the domain of Accessibility.

**Comment by allella** [comment]
This is a UX / usability note, and not directly related to accessibility. Though, I'm imagining accessibility and UX will be reviewed before MVP, and continuously afterwards, so I wanted to leave a breadcrumb to things we've already discussed.

Here are [earlier conversations](https://github.com/freeCodeCamp/chapter/issues/274#issuecomment-562442271) about [switches vs checkboxes](https://uxdesign.cc/ui-cheat-sheet-dropdown-field-a30025c0f432#521e) and [when to use dropdowns vs other things](https://uxdesign.cc/ui-cheat-sheet-dropdown-field-a30025c0f432#521e).

**Comment by Chirag2193** [comment]
I would like to contribute to improve accessibility. Is there anything that I can pick up?

**Comment by allella** [comment]
@Chirag2193 yes, we don't have specific tasks, but there is a list of best practices and links to resources at the top of this thread.

We have not directly focused on the accessibility of the [development copy of the site](https://chapter.freecodecamp.org/), so any small pull requests that address the best practices would be appreciated.

**Comment by AEDesigns** [comment]
We could use tools like Andi, axe and/or Wave to check the sites current
accessibility holes and put together a to-do list

**Comment by Chirag2193** [comment]
Issue: Alternative Text for Images

I see that images associated with events do not yet have alternative texts. I noticed that when creating an event, we ask for the image url, but it would be a good idea to also ask for alternate text. If the user does not give an alternative text, we can set a default text for it.

As an MVP, it would be excellent to include a default alternative text similar to 'Image describing the event' and eventually make it configurable so that the image uploader can add context to the image via the alternative text field.

**Comment by Chirag2193** [comment]
Issue: Buttons don't meet color contrast WCAG expectations

Currently, when running tests with the axe extension tool. The buttons do not appear to fulfil the colour contrast ratio. However, I don't consider this a problem because it's not far from 4.15 and is currently 4.02. Thoughts?

<img width="516" alt="image" src="https://user-images.githubusercontent.com/20030677/235367344-652c224b-dc06-448b-8946-2bf4cb4364c3.png">

<img width="516" alt="image" src="https://user-images.githubusercontent.com/20030677/235367333-f0ade049-59d0-4556-80ca-49f83c9f7266.png">

**Comment by allella** [comment]
It looks like the button is --chakra-colors-blue-500  and there are other, darker color variables like --chakra-colors-blue-600, --chakra-colors-blue-700.  I don't think anybody will be upset if we move the buttons to the next level or two of darkeness to see if that fixes the contrast ratio.

I agree with default alt text, like

_Event Thumbnail for {events.name}_  on routes like https://chapter.freecodecamp.org/chapters/# 
_Event Banner for {events.name}_ on routes like https://chapter.freecodecamp.org/events/#

**Comment by AEDesigns** [comment]
I'm honestly big against vague alt txt like that because it needs to be
descriptive if its important and empty (alt or alt="") if its decorative.
Event Banner or Event Thumbnail will confuse AT users.

Kind Regards,
Robert Elliott  CEO
P +1 (615) 669 5552
W aedevdesigns.com

31

**Comment by allella** [comment]
Well, I think we purposely avoided images on other pages, like the Chapter page, because it was unclear if they were adding a lot of value. Though, having pages without images often comes off as stark when compared to a typical website.

I'd wager most of these images used by organizers are going to be mildly representative of an event. For example, just on the first set of events on Meetup you find things like a photo of Tacos with an Alt text of _"Social Event at White Duck Taco Shop on Airport Rd"_ and I'd imagine Meetup doesn't give the ability to set alt text.

We decidedly don't want to be like Meetup, or base our decisions on what they do, but in a case like this, is it better to expect an organizer to put "Three Tacos on a Wooden Tray" as alt text for a photo, or just exclude the text?

Here is an image of Meetup.com showing a photo of three tacos and, below that, the underlying browser HTML source code showing a generic alt text that does not describe the actual contents of the photo.
![image](https://user-images.githubusercontent.com/1777776/235490958-32e3a3d2-c2ca-47f9-ab6e-b1cf78405de0.png)
